### PR TITLE
fix: [CI-7925]: Fix for CI queue executions

### PR DESCRIPTION
**Pull request by raghavharness**
## Describe your changes
Fixes for CI queue executions
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53347)

### CODE REVIEW

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java [hunk 1]
```diff
@@ -33,37 +39,62 @@ public class CIBuildEnforcerImpl implements CIBuildEnforcer {
...
+    long macExecutionsCount = queueExecutionUtils.getActiveMacExecutionsCount(
+        accountID, List.of(Status.QUEUED.toString(), Status.RUNNING.toString()));
+    long currExecutionCountNonMac = activeExecutionsCount - macExecutionsCount;
+    OSType osType = getOSType(infrastructure);
+    ExecutionLimitSpec executionLimitSpec = getExecutionLimit(accountID);
+    log.info("queue limits for account: {}, total: {}, mac: {}. Current count: total: {}, mac: {}", accountID,
+        executionLimitSpec.getDefaultTotalExecutionCount(), executionLimitSpec.getDefaultMacExecutionCount(),
+        currExecutionCountNonMac, macExecutionsCount);
+    if (osType == OSType.MacOS) {
+      return macExecutionsCount > executionLimitSpec.getDefaultMacExecutionCount();
```
Comment (devkimittal) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java:57]: this should be >=, right? let's 5 running and 5 is count, upcoming task should be queued.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java [hunk 2]
```diff
@@ -224,12 +225,16 @@ public AsyncExecutableResponse executeAsyncAfterRbac(
                                                .stepParameters(RecastOrchestrationUtils.toJson(stepParameters))
                                                .build());
 
-    boolean availableCapacity = buildEnforcer.checkBuildEnforcement(
-        AmbianceUtils.getAccountId(ambiance), Arrays.asList(Status.RUNNING.toString(), Status.QUEUED.toString()));
+    boolean shouldQueue = false;
+    boolean queueConcurrencyEnabled = infrastructure.getType() == Infrastructure.Type.HOSTED_VM
+        && ciFeatureFlagService.isEnabled(QUEUE_CI_EXECUTIONS_CONCURRENCY, AmbianceUtils.getAccountId(ambiance));
```
Comment (devkimittal) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java:230]: we already have account id in 214 line

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java [hunk 3]
```diff
@@ -224,12 +225,16 @@ public AsyncExecutableResponse executeAsyncAfterRbac(
...
+
+    // only check if queue is enabled
+    if (queueConcurrencyEnabled) {
+      shouldQueue = buildEnforcer.shouldQueue(AmbianceUtils.getAccountId(ambiance), infrastructure);
+    }
 
-    boolean queueConcurrencyEnabled =
-        ciFeatureFlagService.isEnabled(QUEUE_CI_EXECUTIONS_CONCURRENCY, AmbianceUtils.getAccountId(ambiance));
-    if (queueConcurrencyEnabled && !availableCapacity) {
+    if (queueConcurrencyEnabled && shouldQueue) {
```
Comment (devkimittal) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java:237]: only shouldQueue is required here I think

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java (hunk 1 above)
Comment (raghavharness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java:57]: No we add to the `ciExecutionMetadata` document first and then calculate it.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java (hunk 3 above)
Comment (raghavharness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/V1/InitializeTaskStepV2.java:237]: yeah

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java [hunk 4]
```diff
@@ -33,37 +39,62 @@ public class CIBuildEnforcerImpl implements CIBuildEnforcer {
   @Inject private QueueExecutionUtils queueExecutionUtils;
   @Inject private ExecutionLimits executionLimits;
   @Inject private ExecutionQueueLimitRepository executionQueueLimitRepository;
+  private static final int MAX_LIMIT = 100000;
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java:42]: Let us give some real number here.
Give it 100 for now.

Comment (raghavharness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/enforcement/CIBuildEnforcerImpl.java]: Done

**Comment by raghavharness** [comment]
trigger uts

**Review by harshjain12 — APPROVED** [review]

**Comment by raghavharness** [comment]
trigger ut6

**Pull request merged**